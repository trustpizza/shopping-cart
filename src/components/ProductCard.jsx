import { useState } from "react";
import PropTypes from "prop-types";

const ProductCard = ({ title, image, price, onClick }) => {
  const [quantity, setQuantity] = useState(1);

  return (
    <div className="card w-full max-w-xs bg-base-100 shadow-lg p-4">
      <img src={image} alt={title} className="w-full h-40 object-cover mb-4 rounded-lg" />
      <h5 className="text-xl font-semibold mb-2">{title}</h5>
      <span className="text-lg font-bold text-primary mb-2">Price: ${price}</span>
      
      <div className="flex items-center gap-2 mb-4">
        <input
          type="number"
          min="1"
          value={quantity}
          onChange={(e) => setQuantity(Number(e.target.value))}
          aria-label="Quantity"
          className="input input-bordered w-16"
        />
        <span className="text-sm">Qty</span>
      </div>
      
      <button
        type="button"
        onClick={() => onClick(quantity)}
        className="btn btn-primary w-full"
      >
        Add to Cart
      </button>
    </div>
  );
};

ProductCard.propTypes = {
  title: PropTypes.string.isRequired,  // Ensures title is a required string
  image: PropTypes.string.isRequired,  // Ensures image is a required string (URL)
  price: PropTypes.number.isRequired,  // Ensures price is a required number
  onClick: PropTypes.func.isRequired,  // Ensures onClick is a required function
};

export default ProductCard;
