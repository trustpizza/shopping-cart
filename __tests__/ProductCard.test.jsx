import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import ProductCard from "../src/components/ProductCard";

describe("ProductCard Component", () => {
  const mockProduct = {
    image: "https://via.placeholder.com/150",
    title: "Sample Product",
    price: "$19.99",
  };

  it("snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ProductCard
          title={mockProduct.title}
          image={mockProduct.image}
          price={mockProduct.price}
          onClick={() => {}}
        />
      </MemoryRouter>
    );

    // Verify that the rendered ProductCard matches the snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders product cards correctly", () => {
    render(
      <MemoryRouter>
        <ProductCard
          title={mockProduct.title}
          image={mockProduct.image}
          price={mockProduct.price}
          onClick={() => {}}
        />
      </MemoryRouter>
    );

    // Check for Product Card image
    const image = screen.getByRole("img");
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", mockProduct.image);

    // Check for Product Card title text
    expect(screen.getByText(mockProduct.title)).toBeInTheDocument();

    // Check for Product Card price text
    expect(screen.getByText(/Price: \$/)).toBeInTheDocument();
    // Check for Product Card quantity input
    const input = screen.getByRole("spinbutton", { name: /quantity/i });
    expect(input).toBeInTheDocument();
    expect(input).toHaveValue(1);

    // Check for Add to Cart button
    const addToCartButton = screen.getByRole("button", { name: /add to cart/i });
    expect(addToCartButton).toBeInTheDocument();
  });

  it("calls onClick function when called", async () => {
    const onClick = vi.fn(); // Mock the onClick function
    const user = userEvent.setup();

    render(
      <MemoryRouter>
        <ProductCard
          title={mockProduct.title}
          image={mockProduct.image}
          price={mockProduct.price}
          onClick={onClick}
        />
      </MemoryRouter>
    );

    // Check that the button exists
    const addToCartButton = screen.getByRole("button", { name: /add to cart/i });
      // Simulate Add to Cart button click
    await user.click(addToCartButton);

    // Ensure onClick is called
    expect(onClick).toHaveBeenCalledTimes(1);
  })

  it("updates quantity input value when changed", async () => {
    render(
      <MemoryRouter>
        <ProductCard
          title={mockProduct.title}
          image={mockProduct.image}
          price={mockProduct.price}
          onClick={() => {}}
        />
      </MemoryRouter>
    );

    const user = userEvent.setup();
    const qtySelector = screen.getByRole("spinbutton", { name: "Quantity" })
    // Default value is 1
    expect(qtySelector).toHaveValue(1);
    
    await userEvent.clear(qtySelector);
    await userEvent.type(qtySelector, "3");

    expect(qtySelector).toHaveValue(3);
  });
});
