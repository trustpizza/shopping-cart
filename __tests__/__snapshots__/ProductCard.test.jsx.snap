// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard Component > snapshot 1`] = `
<DocumentFragment>
  <div
    class="card w-full max-w-xs bg-base-100 shadow-lg p-4"
  >
    <img
      alt="Sample Product"
      class="w-full h-40 object-cover mb-4 rounded-lg"
      src="https://via.placeholder.com/150"
    />
    <h5
      class="text-xl font-semibold mb-2"
    >
      Sample Product
    </h5>
    <span
      class="text-lg font-bold text-primary mb-2"
    >
      Price: $$19.99
    </span>
    <div
      class="flex items-center gap-2 mb-4"
    >
      <input
        aria-label="Quantity"
        class="input input-bordered w-16"
        min="1"
        type="number"
        value="1"
      />
      <span
        class="text-sm"
      >
        Qty
      </span>
    </div>
    <button
      class="btn btn-primary w-full"
      data-testid="add-to-cart"
      type="button"
    >
      Add to Cart
    </button>
  </div>
</DocumentFragment>
`;
